assets/src/edit-story/app/history/test/reducer.js

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies
+ */
+import {
+  default as useReducer,
+  SET_CURRENT_STATE,
+  CLEAR_HISTORY,
+  EMPTY_STATE,
+  REPLAY,
+} from '../reducer';
+
+describe('reducer', () => {
+  const size = 5;
+  const reducer = useReducer(size);
+  it('should throw error if unknown action given', () => {
+    expect(() =>
+      reducer(EMPTY_STATE, {
+        type: 'UNKNOWN_ACTION',
+        payload: {},
+      })
+    ).toThrow(/Unknown history reducer action: UNKNOWN_ACTION/i);
+  });
+
+  describe('addEntry', () => {
+    it('should add entry to the beginning of the history', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 1 }],
+      };
+
+      const newEntry = { id: 2 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.entries[0]).toStrictEqual(newEntry);
+    });
+
+    it('should clear undone states when adding a new entry', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 1 }, { id: 2 }, { id: 3 }],
+        offset: 2,
+      };
+      const newEntry = { id: 4 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.entries).toHaveLength(2);
+      expect(result.entries[1].id).toStrictEqual(3);
+      expect(result.entries[0].id).toStrictEqual(4);
+    });
+
+    it('should set correct state values when adding new entry', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        versionNumber: 3,
+        entries: [{ id: 1 }],
+        offset: 2,
+      };
+      const newEntry = { id: 2 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.versionNumber).toStrictEqual(4);
```

Comment:
I think we just want to check, that it's greater than 3 - we don't really care how much, as long as it's ever-increasing.

Reply:
Doesn't harm to check for exact version though -- e.g. to ensure it's not running more than once for some reason in the future. Thoughts?

Reply:
Well, I suppose that this test would not discover it anyway ;)

Anyway -- since it's expected to increase by exactly 1 in this case then probably doesn't harm to test exactly that, too.